init.lua

```diff
@@ -30,6 +30,20 @@ local function destroy(handle)
 end
 ffi.gc(cudnn.handle, destroy)
 
+cudnn.typemap = {
+   ['torch.CudaHalfTensor']   = 'CUDNN_DATA_HALF',
+   ['torch.CudaTensor']       = 'CUDNN_DATA_FLOAT',
+   ['torch.CudaDoubleTensor'] = 'CUDNN_DATA_DOUBLE',
+}
+
+-- TODO: determine if device supports true half and use true half on it
```

Comment:
OK I was able to run the test on Tesla card. 
Looks like some sizes (and we use random on the test) may not be supported for backward methods, I will get more info on that. What seem to be straightforward is that backward descriptors should be allocated lazily only if they are actually needed:


Running test on device: #2 : Tesla P100-SXM2-16GB

Testing torch.CudaHalfTensor
Running 37 tests
 1/37 VolumetricMaxPooling_batch ........................................ [PASS]
 2/37 SpatialMaxPooling_single .......................................... [PASS]
 3/37 SpatialCrossEntropyCriterion ...................................... [PASS]
 4/37 SpatialMaxPooling_batch ........................................... [PASS]
 5/37 SpatialAveragePooling_batch ....................................... [PASS]
 6/37 LogSoftMax_single ................................................. [PASS]
 7/37 functional_bias2D ................................................. [PASS]
 8/37 SpatialBatchNormalization ......................................... [PASS]
 9/37 SpatialConvolution_forward_single ................................. [ERROR]
10/37 VolumetricMaxPooling_single ....................................... [PASS]
11/37 SpatialFullConvolution_backward_batch ............................. [ERROR]
12/37 TemporalConvolution_padding_batch ................................. [ERROR]
13/37 SpatialConvolution_backward_batch ................................. [ERROR]
14/37 Tanh_single ....................................................... [PASS]
15/37 ClippedReLU_single ................................................ [PASS]
16/37 ReLU_single ....................................................... [PASS]
17/37 Sigmoid_single .................................................... [PASS]
18/37 VolumetricConvolution_forward_single .............................. [ERROR]
19/37 Tanh_batch ........................................................ [PASS]
20/37 SoftMax_batch ..................................................... [PASS]
21/37 SpatialCrossMapLRN_batch .......................................... [PASS]
22/37 VolumetricBatchNormalization ...................................... [PASS]
23/37 SpatialConvolution_backward_single ................................ [ERROR]
24/37 ReLU_batch ........................................................ [PASS]
25/37 TemporalConvolution_single ........................................ [ERROR]
26/37 TemporalConvolution_batch ......................................... [ERROR]
27/37 Sigmoid_batch ..................................................... [PASS]
28/37 SpatialLogSoftMax ................................................. [PASS]
29/37 TemporalConvolution_reduceBatchSize ............................... [ERROR]
30/37 LogSoftMax_batch .................................................. [PASS]
31/37 BatchNormalization ................................................ [PASS]
32/37 SpatialFullConvolution_forward_batch .............................. [ERROR]
33/37 functional_maxpooling2d ........................................... [PASS]
34/37 SpatialConvolution_forward_batch .................................. [ERROR]
35/37 SpatialAveragePooling_single ...................................... [PASS]
36/37 VolumetricConvolution_backward_single ............................. [ERROR]
37/37 SoftMax_single .................................................... [PASS]
Completed 173 asserts in 37 tests with 0 failures and 12 errors
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
SpatialConvolution_forward_single
 Function call failed
/usr/share/lua/5.1/cudnn/init.lua:88: Error in CuDNN: CUDNN_STATUS_NOT_SUPPORTED (cudnnGetConvolutionBackwardDataWorkspaceSize)
stack traceback:
        [C]: in function 'error'
        /usr/share/lua/5.1/cudnn/init.lua:88: in function 'errcheck'
        /usr/share/lua/5.1/cudnn/SpatialConvolution.lua:313: in function 'createIODescriptors'
        /usr/share/lua/5.1/cudnn/SpatialConvolution.lua:364: in function 'forward'
        test/test.lua:176: in function 'test'
        test/test.lua:189: in function <test/test.lua:155>
        [C]: in function 'xpcall'
        /usr/share/lua/5.1/torch/Tester.lua:477: in function '_pcall'
        /usr/share/lua/5.1/torch/Tester.lua:436: in function '_run'
        /usr/share/lua/5.1/torch/Tester.lua:355: in function 'run'
        test/test.lua:1500: in main chunk
        [C]: in function 'dofile'
        /usr/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00406670

--------------------------------------------------------------------------------
SpatialFullConvolution_backward_batch
 Function call failed
/usr/share/lua/5.1/cudnn/init.lua:88: Error in CuDNN: CUDNN_STATUS_NOT_SUPPORTED (cudnnGetConvolutionBackwardDataWorkspaceSize)
stack traceback:
        [C]: in function 'error'
        /usr/share/lua/5.1/cudnn/init.lua:88: in function 'errcheck'
        /usr/share/lua/5.1/cudnn/SpatialFullConvolution.lua:283: in function 'createIODescriptors'
        /usr/share/lua/5.1/cudnn/SpatialFullConvolution.lua:310: in function 'forward'